File: C++/Sorting techniques/Recursive_mergesort.py
def Merge_sort(list1):
    if len(list1)>1:
        middle = len(list1)//2
        left_list = list1[:middle]
        right_list = list1[middle:]

        Merge_sort(left_list)
        Merge_sort(right_list)
        ###################################

        i=0
        j=0
        k=0

        middle = len(left_list)
        high = len(right_list)
        
        while (i<middle and j<high):
                if (left_list[i] < right_list[j]):
                    list1[k] = left_list[i]
                    i+=1
                else:
                    list1[k] = right_list[j]
                    j+=1
                k+=1

        while i < middle:
                    list1[k] = left_list[i]
                    i += 1
                    k += 1

        while j < high:
                    list1[k] = right_list[j]
                    j += 1
                    k += 1


if __name__ == '__main__':
    global list1
    list1 = [8,2,9,6,5,3,7,4]
    print("Before merge_sort : ",*list1)
    Merge_sort(list1)
    print("After merge_sort  : ",*list1)

File: C++/Sorting techniques/Recursive_mergesort.c
#include<stdio.h>
#include<conio.h>

void Display(int A[],int n)
{
	int i;
	for(i=0;i<=n;i++)
	{
		printf("%d ",A[i]);
	}
	printf("\n");
}

void Merge(int A[],int low,int middle,int high)
{
	int	i = low;
	int j = middle+1;
	int k = low;
	int B[100];//int B[high+1];

	// Comparing the elements and putting into the 2nd array
	while(i<=middle && j<=high)
	{
		if(A[i]<A[j])
			B[k++] = A[i++];
		else
			B[k++] = A[j++];
	}

	// if one array completes traversing, the other array elements can be directly copied into the main merged sorted array
	for(;i<=middle;i++)
		B[k++] = A[i];
	for(;j<=high;j++)
		B[k++] = A[j];

	// At-last copy the second array which is the merged and sorted into the 1st array
	for(i=low;i<=high;i++)	
		A[i] = B[i];
}

void Recursive_Merge_Sort(int A[],int low,int high) {
	int middle;
	if (low<high) {
	middle = (low+high)/2;
	Recursive_Merge_Sort(A,low,middle);
	Recursive_Merge_Sort(A,middle+1,high);
	Merge(A,low,middle,high);
	}
}

int main()
{
	int A[] = {8,2,9,6,5,3,7,4},n=7;
	
	printf("Array before sorting : ");
	Display(A,n-1);

	//Iterative_Merge_Sort(A,n);
	Recursive_Merge_Sort(A,0,n);

	printf("Array after sorting  : ");
	Display(A,n-1);

	return 0;
}